app/models/project.rb

```diff
@@ -145,29 +145,47 @@ def add_observation(obs)
     imgs.each { |img| images.push(img) }
   end
 
+  # Nimmo note: I remember this one was a lot of work to get right.
+  # IIRC, the Arel was definitely more efficient than the AR for this join.
+  # But it's been so long, i honestly don't remember. Viva la small PR!
+
   # Remove observation (and its images) from this project. Saves it.
   def remove_observation(obs)
     return unless observations.include?(obs)
 
     imgs = obs.images.select { |img| img.user_id == obs.user_id }
+
+    # Leave images which are attached to other observations
+    # still attached to this project.
     if imgs.any?
-      img_ids = imgs.map(&:id).map(&:to_s).join(",")
-      # Leave images which are attached to other observations
-      # still attached to this project.
-      leave_these_img_ids = Image.connection.select_values(%(
-        SELECT io.image_id FROM images_observations io, observations_projects op
-        WHERE io.image_id IN (#{img_ids})
-          AND io.observation_id != #{obs.id}
-          AND io.observation_id = op.observation_id
-          AND op.project_id = #{id}
-      )).map(&:to_i)
+      leave_these_img_ids = Image.connection.select_values(
+        arel_select_leave_these_img_ids(obs, imgs).to_sql
+      ).map(&:to_i)
+
       imgs.reject! { |img| leave_these_img_ids.include?(img.id) }
     end
+
     observations.delete(obs)
     imgs.each { |img| images.delete(img) }
     update_attribute(:updated_at, Time.zone.now)
   end
```

Comment:
```suggestion
  def remove_observation(obs)
    return unless observations.include?(obs)

    imgs_to_delete(obs).each { |img| images.delete(img) }
    observations.delete(obs)
    update_attribute(:updated_at, Time.zone.now)
  end

  def imgs_to_delete(obs)
    imgs = obs.images.select { |img| img.user_id == obs.user_id }
    return imgs if imgs.none?

    # Do not delete images which are attached to other observations
    # still attached to this project.
    leave_these_img_ids = Image.connection.select_values(
      arel_select_leave_these_img_ids(obs, imgs).to_sql
    ).map(&:to_i)

    imgs.reject { |img| leave_these_img_ids.include?(img.id) }
  end
```
Gets rid of a hidden ABC violation

Reply:
Thanks Joe - this is definitely an improvement.

Reply:
Committed Joe's suggestion. I feel like this PR is ready to merge...?